main.js

```diff
@@ -81,7 +116,123 @@ var createChart = function createChart (element, data, opts) {
     return (bar.label === label);
   }
 
+  function enableDragging (selectedData) {
+    function byLabel (label) {
+      return function (obj) {
+        return obj.label === label;
+      };
+    }
+
+    function onDragLeft (d) {
+      var currentX = timeScale(new Date(d.startedAt * 1000));
+      var newX = currentX + d3.event.dx;
+      var newTime = timeScale.invert(newX).getTime() / 1000;
+
+      if (newTime >= d.endedAt) { return; }
+
+      var siblingsToTheLeft = data.filter(byLabel(d.label)).filter(function (sibling) {
+        return sibling.startedAt <= d.startedAt && sibling != d;
+      });
+
+      var startedAtBoundary = siblingsToTheLeft.map(function (s) {
+        return s.endedAt;
+      }).sort().pop();
+
+      if (startedAtBoundary && newTime <= startedAtBoundary) { return; }
+
+      d.startedAt = newTime;
+
+      d3.select('rect.selected')
+        .attr('x', newX)
+        .attr('width', computeBarWidth);
+
+      d3.select('rect#dragLeft')
+        .attr('x', newX - (dragBarSize / 2));
+
+      opts.onBarChanged(d);
+    }
+
+    function onDragRight (d) {
+      var currentX = timeScale(new Date(d.endedAt * 1000));
+      var newX = currentX + d3.event.dx;
+      var newTime = timeScale.invert(newX).getTime() / 1000;
+
+      if (newTime <= d.startedAt) { return; }
+
+      var siblingsToTheRight = data.filter(byLabel(d.label)).filter(function (sibling) {
+        return sibling.startedAt >= d.endedAt && sibling != d;
+      });
+
+      var endedAtBoundary = siblingsToTheRight.map(function (s) {
+        return s.startedAt;
+      }).sort().shift();
+
+      if (endedAtBoundary && newTime >= endedAtBoundary) { return; }
+
+      d.endedAt = newTime;
+
+      d3.select('rect.selected')
+        .attr('width', computeBarWidth);
+
+      d3.select('rect#dragRight')
+        .attr('x', newX - (dragBarSize / 2));
+
+      opts.onBarChanged(d);
+    }
```

Comment:
These `onDragRight` and `onDragLeft` functions are doing very similar things. The first three lines and last of each are identical, and the code in between is very similar (though not similar enough to refactor away IMO)

It might be time to start giving names to groups of operations by extracting them out into their own functions, even if they're only going to be used once. Something like:

```
function onDragRight (bar) {
      var newValue = getNewValue(bar); // Or heck, just create a whole new "bar"
      // var newBar = buildBar(bar, d3);

      // Maybe combine the two guards...
      if (newTime.x <= bar.startedAt) { return; }
      if (overlapWithFutureBar(bar, newValue)) { return; }

      bar.endedAt = newValue.time;

      // arguable whether these should be pulled out or not.
      d3.select('rect.selected')
        .attr('width', computeBarWidth);

      d3.select('rect#dragRight')
        .attr('x', newValue.x - (dragBarSize / 2));

      opts.onBarChanged(bar);
    }
```

It might also be time (in a different PR) to start pulling some behaviour out of `main.js`.